File: HTTPShortcuts/curl_command/src/main/java/ch/rmy/curlcommand/CommandLineBuilder.java
package ch.rmy.curlcommand;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;

public class CommandLineBuilder {

    private static final String OPTION_PATTERN = "^-{1,2}[a-zA-Z0-9][a-zA-Z0-9-_]*$";

    private final String command;
    private final List<String> arguments = new ArrayList<>();

    public CommandLineBuilder(String command) {
        this.command = command;
    }

    public CommandLineBuilder argument(Object argument) {
        arguments.add(escapeIfNecessary(argument.toString()));
        return this;
    }

    private String escapeIfNecessary(String string) {
        return needsEscaping(string) ? escape(string) : string;
    }

    private boolean needsEscaping(String string) {
        return string.contains("\"") || string.contains(" ");
    }

    private String escape(String string) {
        return "\"" + string.replaceAll("\\\\", Matcher.quoteReplacement("\\")).replaceAll("\"", Matcher.quoteReplacement("\\\"")) + "\"";
    }

    public CommandLineBuilder option(String option, Object... arguments) {
        if (!option.matches(OPTION_PATTERN)) {
            throw new IllegalArgumentException();
        }
        this.arguments.add(option);
        for (Object argument : arguments) {
            argument(argument);
        }
        return this;
    }

    public String build() {
        StringBuilder stringBuilder = new StringBuilder();
        stringBuilder.append(command);
        for (String argument : arguments) {
            stringBuilder.append(' ');
            stringBuilder.append(argument);
        }
        return stringBuilder.toString();
    }

}

File: HTTPShortcuts/curl_command/src/main/kotlin/ch/rmy/curlcommand/CommandLineBuilder.kt
package ch.rmy.curlcommand

import java.util.regex.Matcher

class CommandLineBuilder(private val command: String) {

    private val arguments = mutableListOf<String>()

    fun argument(argument: Any) = also {
        arguments.add(escapeIfNecessary(argument.toString()))
    }

    fun option(option: String, vararg arguments: Any) = also {
        if (!option.matches(OPTION_PATTERN.toRegex())) {
            throw IllegalArgumentException()
        }
        this.arguments.add(option)
        for (argument in arguments) {
            argument(argument)
        }
    }

    fun build(): String {
        val stringBuilder = StringBuilder()
        stringBuilder.append(command)
        for (argument in arguments) {
            stringBuilder.append(' ')
            stringBuilder.append(argument)
        }
        return stringBuilder.toString()
    }

    companion object {

        private const val OPTION_PATTERN = "^-{1,2}[a-zA-Z0-9][a-zA-Z0-9-_]*$"

        internal fun escapeIfNecessary(string: String): String =
            if (needsEscaping(string)) escape(string) else string

        private fun needsEscaping(string: String): Boolean =
            string.contains("\"") || string.contains(" ") || string.contains("<") || string.contains(">")

        private fun escape(string: String): String =
            "\"" + string.replace("\\\\".toRegex(), Matcher.quoteReplacement("\\")).replace("\"".toRegex(), Matcher.quoteReplacement("\\\"")) + "\""
    }
}
